### PR TITLE
Freeze subapp on adding to parent

### PR DESCRIPTION
**Pull request by asvetlov**
Fix for #2656

### CODE REVIEW

File: aiohttp/web_app.py [hunk 1]
```diff
@@ -163,6 +163,7 @@ def add_subapp(self, prefix, subapp):
         self.router.register_resource(resource)
         self._reg_subapp_signals(subapp)
         self._subapps.append(subapp)
+        subapp.freeze()
```
Comment (pfreixes) [inline aiohttp/web_app.py:166]: Even though is not part of that PR I'm wondering if makes sense does not allow add sub apps that have been frozen before. There is an explicit check for that few lines above.

Comment (asvetlov) [reply aiohttp/web_app.py:166]: What is your use case?
Sharing sub-app instance between different parents / prefixes can open a can of worms.
The restriction was exist from very beginning IIRC

Comment (pfreixes) [reply aiohttp/web_app.py:166]: Just trying to understand the why of this check. Looks like an implicit way to check that the sub app has been already added as a child of one app (indeed could be the same).

Comment (asvetlov) [reply aiohttp/web_app.py:166]: Yes, subapp should be used only once. If you need reuse -- please create another subapp instance.

**Comment by pfreixes** [comment]
No `CHANGES`?

**Comment by webknjaz** [comment]
If the app is one-shot instance, let's store its prefix there.

**Comment by asvetlov** [comment]
Forgot to push already committed changes.

**Comment by asvetlov** [comment]
> If the app is one-shot instance, let's store its prefix there.

Sorry, I don't follow.
Application prefix is stored in `PrefixedSubAppResource` (the resource is required anyway).

**Comment by webknjaz** [comment]
So how do I access `app.prefix`?

**Comment by asvetlov** [comment]
What is your use case? There is no such attribute as `handler.path`, why do you need `app.prefix`?

**Comment by pfreixes** [comment]
Feel free to merge when you consider that the conversation is resolved.

**Comment by webknjaz** [comment]
@asvetlov I've got swagger package behaving strange with base url. I have a hope to try improving it by analyzing subapp prefixes.

**Comment by asvetlov** [comment]
Do you mean aiohttp-swagger?
If the library has a bug -- it should be fixed.
I doubt if exposing subapp prefix in public API is the best way to do it.

Please open a new issue if needed

**Pull request merged**